### PR TITLE
Migrate h64 to BigInts, add h32 and h64 streaming APIs, plus performance improvements

### LINKED ISSUE #9: Expose a streaming api?
Any chance you'd consider exposing an api that'll enable hashing the content of a large file (that won't fit in memory)? Maybe the equivalent of the `XXH64_update` C api as used in [the streaming variant in this example here](https://github.com/Cyan4973/xxHash#example)?

Fantastic lib btw, it's a little shocking how fast this runs! 🚀

### PR DESCRIPTION
**Pull request by marcusdarmstrong**
Resolves #9 and #25.

This PR includes changes to significantly improve the performance of the existing APIs (mostly as discussed in #25), as well as changes to migrate the h64 APIs to use BigInts and add new streaming APIs for both h32 and h64.

### Open Questions

I'll start with my "open questions"... I've made choices here for purposes of posting the PR, but I'd describe them as the most shaky portion of the changeset, and accordingly warrant some discussion.
1. **`string` behavior**: I've maintained the original API's type contract of returning strings from the `h32` and `h64` functions, however, I don't feel great about that. The fact that these APIs _take_ strings is a bit orthogonal to the idea that they should return them, and incurs a performance cost in the case where the raw numeric hash values are an appropriate result of a string hash. That said, there's certainly some ergonomics to be concerned with, as most users certainly would expect hex string return values for hashes. I've also zero-padded in the general case here, again to align with my perception of user expectation.
2. **`BigInt`/`number` asymmetry between 32-bit and 64-bit APIs**: Since the 32 bit API can express its seed/value via a `number` I've retained the existing API contracts there, however, there's an argument to be made that we can/should use `BigInt`s for both to provide a completely consistent API between the algorithms.
3. **`Hash<T>.update` accepts `string | Uint8Array`, rather than separate `update`/`updateRaw` APIs**: this matches the node `crypto` API, but diverges from the `h*/h*Raw` dichotomy. Removing the branching might provide a performance improvement, so I'm certainly open to either option.
4. **There's currently no changelog in the package.** I'd be happy to add one as a part of this PR, but if you prefer to manage that information some other way, that's fine of course.
5. **I didn't do the semver bump here**, but obviously that's easy if desired.

### Downsides

1. **Bundle size increase**: The streaming APIs added a significant amount of webassembly to the package, and the JS-size increased as well to support them. Theoretically this could be trimmed down a bit with some work (for example, by removing the ability of the wasm module to hash at an arbitrary point in the linear memory, since it's always invoked from 0), but that seemed like an unwarranted tradeoff to me, and it's doubtful we could ever keep the expanded set of functionality within the original bundlesize spec.
2. **Requires bulk memory operations, BigInts, and `TextEncoder.encodeInto`**: These changes use some new APIs and change the target of the library accordingly. Bulk Memory Operations are available in Chrome >=75, Firefox >= 79, Safari >= 15, and Node >= 12.5 (behind a flag) . WASM BigInt support is available in Chrome >=85, Firefox 78, Safari >= 14.1, and Node >= 15. `TextEncoder.encodeInto` is available in Chrome >= 74, Firefox >= 66, Safari >= 14.1, and Node >= 12.11.

The end result there is that the practical library target is now:
- Chrome >= 85
- Firefox >= 79
- Safari >= 15
- Node >= 15

### Benchmarks

Benchmarks run on a 2019 x64 MacBook Pro running Node 17.3.0. `xxhash-wasm-1` is obviously the version from this PR, while `xxhash-wasm` is 0.4.2.

```
xxhash-wasm#h32 x 850,309 ops/sec ±3.17% (74 runs sampled)
xxhash-wasm#h64 x 726,678 ops/sec ±3.66% (79 runs sampled)
xxhash-wasm#h64Raw x 2,826,606 ops/sec ±0.95% (94 runs sampled)
xxhash-wasm-1#h32 x 3,901,259 ops/sec ±1.22% (92 runs sampled)
xxhash-wasm-1#h64 x 3,459,188 ops/sec ±1.19% (90 runs sampled)
xxhash-wasm-1#h64Raw x 9,977,528 ops/sec ±1.20% (90 runs sampled)
Benchmark 1 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 1,110,898 ops/sec ±4.64% (75 runs sampled)
xxhash-wasm#h64 x 731,355 ops/sec ±3.18% (80 runs sampled)
xxhash-wasm#h64Raw x 1,073,810 ops/sec ±1.50% (87 runs sampled)
xxhash-wasm-1#h32 x 3,931,655 ops/sec ±1.02% (88 runs sampled)
xxhash-wasm-1#h64 x 3,505,964 ops/sec ±1.02% (91 runs sampled)
xxhash-wasm-1#h64Raw x 10,219,203 ops/sec ±0.96% (94 runs sampled)
Benchmark 10 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 846,448 ops/sec ±3.35% (76 runs sampled)
xxhash-wasm#h64 x 586,251 ops/sec ±3.37% (80 runs sampled)
xxhash-wasm#h64Raw x 155,676 ops/sec ±1.15% (92 runs sampled)
xxhash-wasm-1#h32 x 2,056,868 ops/sec ±2.52% (91 runs sampled)
xxhash-wasm-1#h64 x 3,249,041 ops/sec ±1.31% (92 runs sampled)
xxhash-wasm-1#h64Raw x 8,250,841 ops/sec ±2.24% (88 runs sampled)
Benchmark 100 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 642,905 ops/sec ±2.57% (73 runs sampled)
xxhash-wasm#h64 x 604,470 ops/sec ±2.29% (73 runs sampled)
xxhash-wasm#h64Raw x 16,618 ops/sec ±0.93% (95 runs sampled)
xxhash-wasm-1#h32 x 2,323,392 ops/sec ±1.75% (91 runs sampled)
xxhash-wasm-1#h64 x 2,552,064 ops/sec ±0.85% (94 runs sampled)
xxhash-wasm-1#h64Raw x 5,657,834 ops/sec ±1.47% (95 runs sampled)
Benchmark 1000 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 165,309 ops/sec ±1.97% (86 runs sampled)
xxhash-wasm#h64 x 171,561 ops/sec ±2.38% (88 runs sampled)
xxhash-wasm#h64Raw x 1,644 ops/sec ±1.17% (92 runs sampled)
xxhash-wasm-1#h32 x 548,383 ops/sec ±1.03% (94 runs sampled)
xxhash-wasm-1#h64 x 831,613 ops/sec ±1.12% (95 runs sampled)
xxhash-wasm-1#h64Raw x 1,162,121 ops/sec ±0.96% (93 runs sampled)
Benchmark 10000 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 19,001 ops/sec ±3.24% (83 runs sampled)
xxhash-wasm#h64 x 22,422 ops/sec ±1.44% (87 runs sampled)
xxhash-wasm#h64Raw x 166 ops/sec ±1.13% (85 runs sampled)
xxhash-wasm-1#h32 x 60,206 ops/sec ±0.84% (90 runs sampled)
xxhash-wasm-1#h64 x 98,119 ops/sec ±0.75% (94 runs sampled)
xxhash-wasm-1#h64Raw x 118,755 ops/sec ±0.88% (96 runs sampled)
Benchmark 100000 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 1,832 ops/sec ±1.71% (83 runs sampled)
xxhash-wasm#h64 x 2,152 ops/sec ±1.28% (87 runs sampled)
xxhash-wasm#h64Raw x 16.18 ops/sec ±1.53% (44 runs sampled)
xxhash-wasm-1#h32 x 5,665 ops/sec ±0.83% (94 runs sampled)
xxhash-wasm-1#h64 x 8,587 ops/sec ±0.99% (90 runs sampled)
xxhash-wasm-1#h64Raw x 10,593 ops/sec ±1.06% (92 runs sampled)
Benchmark 1000000 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 142 ops/sec ±0.58% (80 runs sampled)
xxhash-wasm#h64 x 156 ops/sec ±0.59% (79 runs sampled)
xxhash-wasm#h64Raw x 0.83 ops/sec ±1.97% (7 runs sampled)
xxhash-wasm-1#h32 x 334 ops/sec ±1.12% (83 runs sampled)
xxhash-wasm-1#h64 x 429 ops/sec ±0.99% (89 runs sampled)
xxhash-wasm-1#h64Raw x 588 ops/sec ±0.77% (86 runs sampled)
Benchmark 10000000 bytes - Fastest is xxhash-wasm-1#h64Raw

xxhash-wasm#h32 x 12.73 ops/sec ±3.03% (36 runs sampled)
xxhash-wasm#h64 x 14.02 ops/sec ±0.49% (37 runs sampled)
xxhash-wasm#h64Raw x 0.03 ops/sec ±0.51% (5 runs sampled)
xxhash-wasm-1#h32 x 28.65 ops/sec ±0.54% (51 runs sampled)
xxhash-wasm-1#h64 x 35.64 ops/sec ±0.50% (62 runs sampled)
xxhash-wasm-1#h64Raw x 44.88 ops/sec ±0.28% (59 runs sampled)
Benchmark 100000000 bytes - Fastest is xxhash-wasm-1#h64Raw
```

### Change Overview

1. `h64` now uses `BigInt`s for both seeds as well as return values from wasm. The seed is now communicated via an argument rather than memory writes. Raw now directly returns a `BigInt`, rather than a `Uint8Array`.
2. The `Uint8Array` providing access to the wasm memory buffer is now reused and only allocated on memory growth.
3. `string` encoding has been moved to use TextEncoder.encodeInto, after a bunch of benchmarking. In #25 I had noted large performance differences between Buffer.from and TextEncoder—it turns out those differences are _mostly_ ARM-specific (my cited benchmark runs in that discussion are from an M1 MacBook Pro). While Buffer.from does still outperform TextEncoder.encode on small inputs, eliminating the second memcpy involved in using those APIs by using TextEncoder.encodeInto to directly encode into the wasm linear memory yields substanial improvements, as documented above. The downside here is that we have to overprovision the wasm memory to account for the risk that we're encoding high-byte-size code points, but should that extra memory use be a consideration for a particular use case, the Raw APIs should provide a reasonable means of handling it.
4.  I've added two new methods to the primary exports: `create32` and `create64`, which allow creation of stateful hashes that can be used for streaming hash applications. These hashers are closed over an xxh state object which we push into the wasm memory whenever we need to update or digest the hash, and read out afterwards. The WAT is a direct port of the C implementation of xxh, done very manually.

In general, the JS code is extremely divergent from the previous version, so it's probably easiest to review as if it were entirely new code.

Please let me know what adjustments you'd like to see and I'll happily get on them ASAP.

### CODE REVIEW

**Comment by marcusdarmstrong** [comment]
> if you pass a number to the 64-bit API is it still handled correctly and just limits the maximum value or are the bytes misinterpreted by the WASM (as in float in memory is interpreted as a long)?

Passing a number as a BigInt seed is a runtime error: `TypeError: Cannot convert 1 to a BigInt`. My preference would be to avoid the conversion, as constructing a new BigInt is an allocation and as such involves some legitimate overhead.

I've added the relevant testcases, removed the no-longer-compatible CI targets, and updated the readme with 1. a streaming example, 2. a mention of the limitations of the string api and 3. a discussion of engine requirements. After some thought, I've also added two new API methods here—feedback of course welcome—`h32String` and `h64String` that provide raw numeric hashes with the performance of the encodeInto API. Leaving an ergonomic default but providing access to the best possible performance case seemed like the best option to me... There's unfortunately no way to emulate the encodeInto performance via the Raw methods.

**Comment by jungomi** [comment]
> > if you pass a number to the 64-bit API is it still handled correctly and just limits the maximum value or are the bytes misinterpreted by the WASM (as in float in memory is interpreted as a long)?
> 
> Passing a number as a BigInt seed is a runtime error: `TypeError: Cannot convert 1 to a BigInt`. My preference would be to avoid the conversion, as constructing a new BigInt is an allocation and as such involves some legitimate overhead.

Ah okay, then it's fine as it is.

The `h32String`/`h64String` methods are definitely a nice solution, but the names are counter-intuitive, because it would kind of be expected to be the other way around. That brings up the question, whether the default methods (`h32`/`h64`) should actually be returning strings or whether the string version is an additional one. In terms of API design it's certainly more elegant to have the default returning numbers and have a convenience method to directly convert it to a string.
If users expect a string output from the default methods (probably mostly due to backwards-compatibility), I think the `h**String` should be renamed to something that better indicates the difference being it returning a number instead of a string.
I'm personally in favour of having the default methods return a number and adding the convenience methods for string outputs, but the only thing I'm worried about if these changes will get caught, because if the TypeScript types are not checked, it could lead to silent misbehaviour (most likely with string concatenation), because there will not be any runtime error from that.
There are other breaking changes and it's a new major version, so I guess it doesn't really matter. What do you think?

Otherwise the rest looks fine to me.

File: README.md [hunk 1]
```diff
@@ -89,6 +91,26 @@ h32(input); // ee563564
...
+
+`xxhash-wasm` supports a `crypto`-like streaming api, useful for avoiding memory consumption when hashing large amounts of data:
+
+```javascript
+const { create32, create64 } = await xxhash();
+
+// 32-bit version
+create32()
+  .update("some data")
+  .update(Uint8Array.from([1, 2, 3]))
```
Comment (jungomi) [inline README.md:105]: A comment that `.update()` expects either a `string` or `Uint8Array` would be helpful for those who aren't familiar with the crypto streaming API and who haven't looked at the signature down below in the API section.

Comment (marcusdarmstrong) [reply README.md:105]: 👍 Updated.

**Comment by marcusdarmstrong** [comment]
> I'm personally in favour of having the default methods return a number and adding the convenience methods for string outputs, but the only thing I'm worried about if these changes will get caught, because if the TypeScript types are not checked, it could lead to silent misbehaviour (most likely with string concatenation), because there will not be any runtime error from that.
> There are other breaking changes and it's a new major version, so I guess it doesn't really matter. What do you think?

I completely agree. I think the risk of such a structural change going unnoticed on a major version update is quite minor, particularly if it's highlighted in the release notes.

As far as naming goes, how about...
```
h64(string[, BigInt]): BigInt
h64Raw(UInt8Array[, BigInt]: BigInt
h64Hex(string[,BigInt]): string
```

**Comment by jungomi** [comment]
`h64Hex` I don't associate Hex with string. I think it's easiest to keep it close to any string conversion naming, either `h64ToString` to suggest the `.toString()`, or maybe `h64AsString`. The fact that it's converted to a zero padded hex representation is the implementation detail, which is what would be expected from a `.toString()` if there were a native hex number type.
I also thought your current implementation with `h64String` was fine, just having them the other way around to match what would be my understanding of the names.

**Comment by marcusdarmstrong** [comment]
Works for me, I'll go with h64ToString().

File: README.md [hunk 2]
```diff
@@ -73,9 +73,9 @@ import xxhash from "xxhash-wasm";
 xxhash().then(hasher => {
   const input = "The string that is being hashed";
   // 32-bit version
-  hasher.h32(input); // ee563564
+  hasher.h32ToString(input); // ee563564
   // 64-bit version
-  hasher.h64(input); // 502b0c5fc4a5704c
+  hasher.h64ToString(input); // 502b0c5fc4a5704c
```
Comment (jungomi) [inline README.md:78]: Put the output in the comments in quotes, e.g. `"ee563564"`, to make it clearer that it's a string.

For the version below where numbers are returned, maybe also add `(decimal representation)` or `(BigInt)` or something along these lines to clarify that those are not hex representations and hence differ from the string values above.

Also the only difference between the two code blocks was the `async`/`await`, that should probably kept in sync and the string version in a separate one, explaining the availability of the `h**ToString` methods as a convenience to get the hexadecimal representation.

File: README.md [hunk 3]
```diff
@@ -242,53 +240,53 @@ different lengths. *Higher is better*
...
-| 10,000 bytes              | 13,010 ops/sec     | 1,452 ops/sec      | 446,415 ops/sec         | ***787,430 ops/sec***   |
-| 100,000 bytes             | 477 ops/sec        | 146 ops/sec        | 58,339 ops/sec          | ***91,895 ops/sec***    |
-| 1,000,000 bytes           | 36.40 ops/sec      | 12.93 ops/sec      | 5,443 ops/sec           | ***8,379 ops/sec***     |
-| 10,000,000 bytes          | 3.12 ops/sec       | 1.19 ops/sec       | 307 ops/sec             | ***417 ops/sec***       |
-| 100,000,000 bytes         | 0.31 ops/sec       | 0.13 ops/sec       | 28.70 ops/sec           | ***35.63 ops/sec***     |
-
-`xxhash-wasm` outperforms `xxhashjs` significantly, the 32-bit is up to 98 times
+| 1 byte                    | 513,517 ops/sec    | 11,896 ops/sec     | 5,752,446 ops/sec       | ***4,438,501 ops/sec*** |
+| 10 bytes                  | 552,133 ops/sec    | 12,953 ops/sec     | 6,240,640 ops/sec       | ***4,855,340 ops/sec*** |
+| 100 bytes                 | 425,277 ops/sec    | 10,838 ops/sec     | 5,470,011 ops/sec       | ***4,314,904 ops/sec*** |
```
Comment (jungomi) [inline README.md:245]: The 32-bit should be bold, not the 64-bit for up to 100 bytes, as it was mentioned in the text below, although the cross-over point for the 64-bit to become faster is almost exactly at 1,000 bytes or very slightly below it (not 100).

**Review by jungomi — COMMENTED** [review]
Last couple of remarks just for the README

File: README.md (hunk 3 above)
Comment (marcusdarmstrong) [reply README.md]: Ah, right. Fixed. My experience was that there was actually a pretty high degree of variability in those small-byte-size benchmarks—I wouldn't be particularly confident in that 100 vs. 1000 measurement, but might as well summarize the data in the table.

File: README.md (hunk 2 above)
Comment (marcusdarmstrong) [reply README.md]: 👍 Good feedback.

**Comment by jungomi** [comment]
Thank you very much for your contributions. I'll be looking to prepare the release so that `v1.0.0` will be published in the next few days.

**Comment by extremeheat** [comment]
Regarding the switch from two numbers to BigInt for 64-bit seeds: my understanding is that BigInts are [much slower for almost all operations](https://github.com/tc39/proposal-bigint/issues/117) done on them--so is it safe to assume that the performance impact here for call overhead has been tested to be negligible?

**Comment by marcusdarmstrong** [comment]
In short, yes. I did detailed performance evaluation for every change included in this PR—in practice, there’s very little in the way of actual operations that need to be performed on a bigint in the relevant code paths: the math is all done on i64 types inside of easy and then boxed to a bigint on the way out. The only actual bigint operation is the u64-izing, which, while it has overhead, is significantly cheaper than the hi/low munging previously in use.

**Pull request merged**